### DISCUSSION TITLE
Potential JS overhead for sites with just 1-2 simple interactivity features

### DISCUSSION THREAD

**Discussion by felixarntz** [WordPress/gutenberg#53039] [Interactivity API]
While I am very excited to see this project take off, one concern that comes to mind is regarding scenarios where this interactivity API may do more harm than good when it comes to performance.

Generally, for sites with a certain level of interactivity, I think this API will be great, as it will make writing the code for interactive patterns much simpler and since it relies on declarative HTML instead of custom JS, it will also notably reduce the JS footprint.

However, there is also the opposite scenario: Several WordPress sites today are quite simple and come with minimum interactivity (consider e.g. a minimum blog site with e.g. just the WordPress default theme). The typical interactive feature that almost _every_ WordPress site has today is toggling a mobile navigation or submenus in a menu - but if those two are the only interactivity in a URL, the custom JS loaded today for that is notably less than what the interactivity API would require: I appreciate how much care has gone into implementing a lightweight runtime. Yet, 33 KB (8 KB runtime + 25 KB vendors) is still a substantial amount of JS, and a theme could instead cover those small interactivity aspects with a small script of maybe 1-2 KB or less.

I don't want to play the devil's advocate here, but for simple WordPress sites with only little interactivity there is a cost to using this API. We can potentially build a more intelligent approach that only loads more granularly the JS that's actually needed, but it may be impossible to realistically come to a solution that beats the custom JS amount for such a simple site.

Maybe this is just about education, and we have to suggest to developers of themes with very basic interactivity to not rely on the interactivity API and instead manually code it. But then again, if a site uses that theme and installs lots of plugins and custom blocks that make use of the runtime, there would be no point in using custom JS instead of the runtime after all.

From a DX perspective, it's definitely desirable for any interactive code in WordPress to use this API. But we also need to think about the performance impact of the JS that the interactivity API itself takes. It's hard to estimate for how many sites using it will make performance better or worse. 🤔 

Anyways, I'm opening this issue primarily to start a conversation about this. I'm curious whether this has been considered so far, hopefully we can find a solid solution together.

**Comment by luisherranz** [comment]
That was precisely one of the reasons to focus so much on the bundle size. As you say, you would probably need at least 2-3 blocks to match its size. But for example, the current code for the Navigation block is around 4kBs (gzipped). The Interactivity API is 8kBs of vendors and 2kBs of runtime. Most (if not all) of the block themes use the Navigation block, so if we migrate the Navigation block, we're only 6kBs off. Additionally, there are plans to introduce more interactivity in other core blocks, such as the option of an [image lightbox](https://github.com/WordPress/gutenberg/pull/49621) similar to the one in Medium.

What is great about the Interactivity API is the way it scales. If you take a look at the [wpmovies.dev](https://wpmovies.dev/) site, you'll see that the custom code on top of the Interactivity API (the `view.js` files) is pretty small, less than 0.5kBs per interactive block:

<img width="600" alt="Screenshot 2023-04-19 at 16 43 27" src="https://user-images.githubusercontent.com/3305402/233116071-2bec4776-9b59-4553-b88a-5e0781d053a5.png">


_Here, the runtime is a bit bigger (3kBs) because it includes the client-side router, but if you don't activate client-side navigations, it should be around 2kBs. And maybe we can strip out another kilobyte from the runtime if we code-split the core directives, and ship only 1kB by default. However, this is still up for discussion as it may also be detrimental to enqueue too many files._

The small size of the `view.js` files remains true even for complex blocks. Look how simple it's the JavaScript necessary to enable client-side form submissions of the comments form: [view.js](https://github.com/WordPress/gutenberg/blob/e5e61a04a39eb6cbf1aa58bc9ae7956c884fedf4/packages/block-library/src/post-comments-form/view.js). That means that after 2-3 interactive blocks (which is a low number), sites should start seeing a very good reduction in their JS bundle.

The only option I see to solve this would be to ship two implementations of each block, one using vanilla JavaScript, and one using the Interactivity API. Then, inspect each page and if there are less than 2-3 interactive blocks, ship the vanilla JavaScript version instead. It's doable, but it kind of defeats the purpose of the enhanced developer experience for a gain that may not be so big (~6kBs on sites with a single interactive block). 

Because there's also the question of _"make it as fast as possible"_ versus _"make it fast enough"_. The [pagespeed ingsights of the wpmovies.dev page](https://pagespeed.web.dev/analysis/https-wpmovies-dev/6ijylga1re?form_factor=mobile) are already 100, and that includes the 10kB runtime, including the client-side router, an aggressive prefetching (for demonstration purposes) and the 3kBs of the 4-5 interactive blocks present in the page.

<img width="400" alt="Screenshot 2023-04-19 at 17 07 54" src="https://user-images.githubusercontent.com/3305402/233123493-734674e8-43dd-4af4-b06b-c5f7ccf4c0fc.png">

Oh, I almost forgot. There's yet another optimization that would affect this and we haven't implemented yet: lazy hydration, the ability to delay the download and execution of an interactive block until the CPU/network is idle or the block is about to enter the viewport. It's certainly doable [and we've experimented with it in the past](https://github.com/WordPress/block-interactivity-experiments/blob/main-custom-elements-hydration/src/gutenberg-packages/wordpress-element.js#L53-L94), although [we haven't added it to the current version yet](https://github.com/WordPress/block-interactivity-experiments/issues/75).

---

All in all, I want to acknowledge that this is solely my perspective, and I may be terribly biased in thinking that the additional cost on sites with only one block is not too high to bear. However, it's possible that others may have a different opinion on this matter 🙂

**Comment by felixarntz** [comment]
Thanks for the detailed reply @luisherranz!

First of all, like I said, I definitely acknowledge that a lot has gone into making the runtime small, I hope it's clear that I'm not pushing back on that :)

I would like to reply to some of your points though:
> But for example, the current code for the Navigation block is around 4kBs (gzipped).

To be fair, by far the most of that is the `view-modal.min.js` file, which I believe is built with Webpack, which makes it way larger than it would have to be. Build tools are great, but when running a simple script with a couple lines through Webpack, you end up with a file notably larger than the original source file. So I would take your statement here with a grain of salt, it depends on how the block script is being developed and whether it needs/uses Webpack or just a simple, more traditional minification build step.

> _And maybe we can strip out another kilobyte from the runtime if we code-split the core directives, and ship only 1kB by default. However, this is still up for discussion as it may also be detrimental to enqueue too many files._

That sounds promising, I'd love to see that. Enqueuing more files as of today isn't really a concern as much as it used to be - I think a bit more code-splitting to allow for more granular loading of the JS would be beneficial!

> The only option I see to solve this would be to ship two implementations of each block, one using vanilla JavaScript, and one using the Interactivity API. Then, inspect each page and if there are less than 2-3 interactive blocks, ship the vanilla JavaScript version instead.

That's an interesting idea! I agree with you it may not be worth the effort, but at least something worth considering :)

Regarding the PageSpeed score of the example page, while that is great, I'd argue this is not an example of the problem I'm concerned about. That page is properly using the Interactivity API and doesn't mix it with other more traditional JS usage. If a site uses e.g. one block that uses the Interactivity API, a few other blocks that use vanilla JS, and a few other blocks that load jQuery (hopefully not 😆), then the Interactivity API would add more overhead (to what of course would already be a problematic page setup in the first place, just to be fair).

Anyway, I really like your idea of breaking apart the runtime further in a way that allows us to load its parts more granularly on demand - of course within reason, but that may get us somewhere :)

**Comment by westonruter** [comment]
> There's yet another optimization that would affect this and we haven't implemented yet: lazy hydration, the ability to delay the download and execution of an interactive block until the CPU/network is idle or the block is about to enter the viewport. It's certainly doable [and we've experimented with it in the past](https://github.com/WordPress/block-interactivity-experiments/blob/main-custom-elements-hydration/src/gutenberg-packages/wordpress-element.js#L53-L94), although [we haven't added it to the current version yet](https://github.com/WordPress/block-interactivity-experiments/issues/75).

I love this idea. Commenting is a prime area that could benefit from this. Today the `comment-reply` script is added to every single page, as a blocking script (but at least it is in the footer). With the impending availability of script loading strategies, such scripts can at least be loaded `async`/`defer` but they'll still be loaded even if the user never even gets to the comments. So if `comment-reply` can be lazy-loaded, then this would be a further 1.3kB reduction in the total initially-loaded JS.
